### PR TITLE
Fix #487: remove dust in stake & share calc

### LINKED ISSUE #487: Khala Network collators stuck at some extrinsics
(More detail will be added)

### PR DESCRIPTION
**Pull request by h4x3rotab**
Fixes #487 

- Handle dust when calculating stake & shares in the following rules:
    - Dust in user's locked stake will be removed by slash
    - Dust in a user's share is removed, and always maintain the invariant: sum(user shares) == pool's total share
    - Dust in a pool's total_stake is removed
    - Dust in withdraw queue is removed
    - Dust in payout report from GK is dismissed
- Add a migration to remove all the existing dust in the storage
    - Analyzed with `try_runtime`, and with state dumped for offline analysis

## Tested

Unit test with grcov:

image

### CODE REVIEW

File: pallets/phala/src/stakepool/migrations.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+		// Remove dust in stakers and collect dust shares that will be removed in stake pools
+		// later.
+		PoolStakers::<T>::translate(
+			|key: (u64, T::AccountId), mut user: UserStakeInfo<T::AccountId, BalanceOf<T>>| {
+				let (pid, account) = key;
+				// Shares
+				let (shares, shares_dust) = extract_dust(user.shares);
+				user.shares = shares;
+				// Accumulate the dust to the table to ensure shares invarant
```
Comment (h4x3rotab) [inline pallets/phala/src/stakepool/migrations.rs:148]:
```suggestion
				// Accumulate the dust to the table to ensure shares invariant
```

File: pallets/phala/src/stakepool.rs [hunk 2]
```diff
@@ -641,13 +672,21 @@ pub mod pallet {
...
-				if pool_info.total_shares == Zero::zero() {
+				if balance_close_to_zero(pool_info.total_shares) {
 					Self::deposit_event(Event::<T>::RewardDismissedNoShare(pool_info.pid, rewards));
 					return;
 				}
 				let commission = pool_info.payout_commission.unwrap_or_default() * rewards;
 				pool_info.owner_reward.saturating_accrue(commission);
-				pool_info.distribute_reward(rewards - commission);
+				let to_distribute = rewards - commission;
+				if is_positive_balance(to_distribute) {
```
Comment (tolak) [inline pallets/phala/src/stakepool.rs:682]: `is_positive_balance` is a little bit confused here, `to_distribute` is actually always a positive value, how about rename it to `not_dust_balance` or something else?

**Comment by tolak** [comment]
Except the function name `is_positive_balance `, others LGTM

File: pallets/phala/src/stakepool.rs (hunk 2 above)
Comment (h4x3rotab) [reply pallets/phala/src/stakepool.rs:682]: Sounds good.

**Pull request merged**